spec/std/software_version_spec.cr

```diff
@@ -0,0 +1,168 @@
+require "spec"
+require "software_version"
+
+private def v(string)
+  SoftwareVersion.parse string
+end
+
+private def v(version : SoftwareVersion)
+  version
+end
+
+# Assert that two versions are equal. Handles strings or
+# SoftwareVersion instances.
+private def assert_version_equal(expected, actual)
+  v(actual).should eq v(expected)
+  v(actual).hash.should eq(v(expected).hash), "since #{actual} == #{expected}, they must have the same hash"
+end
+
+# Refute the assumption that the two versions are equal.
+private def refute_version_equal(unexpected, actual)
+  v(actual).should_not eq v(unexpected)
+  v(actual).hash.should_not eq(v(unexpected).hash), "since #{actual} != #{unexpected}, they must not have the same hash"
+end
+
+describe SoftwareVersion do
+  it ".valid?" do
+    SoftwareVersion.valid?("5.1").should be_true
+    SoftwareVersion.valid?("an invalid version").should be_false
+  end
+
+  it "equals" do
+    assert_version_equal "1.2", "1.2"
+    assert_version_equal "1.2.b1", "1.2.b.1"
+    assert_version_equal "1.0+1234", "1.0+1234"
+    refute_version_equal "1.2", "1.2.0"
+    refute_version_equal "1.2", "1.3"
+  end
+
+  it ".new with number" do
+    assert_version_equal "1", SoftwareVersion.new(1)
+    assert_version_equal "1.0", SoftwareVersion.new(1.0)
+  end
+
+  it ".new with segments" do
+    SoftwareVersion.new(1).to_s.should eq "1"
+    SoftwareVersion.new(1, 0).to_s.should eq "1.0"
+    SoftwareVersion.new(1, 0, 0).to_s.should eq "1.0.0"
+    SoftwareVersion.new(1, 0, 0, prerelease: "rc1").to_s.should eq "1.0.0-rc1"
+    SoftwareVersion.new(1, 0, 0, prerelease: "rc1", metadata: "build8").to_s.should eq "1.0.0-rc1+build8"
+    SoftwareVersion.new(1, prerelease: "rc1").to_s.should eq "1-rc1"
+    SoftwareVersion.new(1, prerelease: "rc1", metadata: "build8").to_s.should eq "1-rc1+build8"
+  end
+
+  describe ".parse" do
+    it "parses metadata" do
+      SoftwareVersion.parse("1.0+1234")
+    end
+
+    it "valid values for 1.0" do
+      {"1.0", "1.0 ", " 1.0 ", "1.0\n", "\n1.0\n", "1.0"}.each do |good|
+        assert_version_equal "1.0", good
+      end
+    end
+
+    it "invalid values" do
+      invalid_versions = %w[
+        junk
+        1.0\n2.0
+        1..2
+        1.2\ 3.4
+      ]
+
+      # DON'T TOUCH THIS WITHOUT CHECKING CVE-2013-4287
+      invalid_versions << "2.3422222.222.222222222.22222.ads0as.dasd0.ddd2222.2.qd3e."
+
+      invalid_versions.each do |invalid|
+        expect_raises ArgumentError, "Malformed version string #{invalid.inspect}" do
+          SoftwareVersion.parse invalid
+        end
+        SoftwareVersion.parse?(invalid).should be_nil
+      end
+    end
+
+    it "empty version" do
+      ["", "   ", " "].each do |empty|
+        SoftwareVersion.parse(empty).to_s.should eq "0"
+      end
+    end
+  end
+
+  it "#<=>" do
+    # This spec has changed from Gems::Version where both where considered equal
+    v("1.0").should be < v("1.0.0")
```

Comment:
I find it quite unintuitive that these are not equal. What is the motivation?

Reply:
They are in fact *different* with a different number of segments. Both interpretations are legitimate and I don't think there is a definitive argument for one or the other. It mostly depends on the context which behaviour would be expected. The practical relevance however is pretty small anyway. You're usually comparing software versions in a specific context (such as a versioned software package) where changes in version format are pretty rare and unlikely to assign a version that could be considered equal twice. So I don't think it matters much what the comparison returns.

When in doubt, I'd usually prefer to stick to the example of `Gem::Version`, but there are implications on hash performance. Equality in `<=>` should also result in equal hashes and ignoring trailing zero segments would add more complexity to the hasher, thus reducing performance.

Reply:
It has been a cause of bugs in shards, and now shards considers 1 == 1.0 == 1.0.0 = ...

Reply:
@ysbaddaden What kind of bugs?